### PR TITLE
MM-34646: force-refresh user tokens

### LINKED ISSUE #431: Access Token Upgrade To New Format
I got an email from Github informing me that the token used for this integration is deprecated.

Is a change to this plugin required to work properly with the new format?

> [GitHub API] Update your access token to the new format
> GitHub <noreply@github.com>
> Mon 4/5/2021 7:05 PM
> Hi @frakman1,
> 
> We noticed that, at April 5th, 2021 at 23:05 (UTC) your application (Mattermost GitHub Plugin - MM, with a user-agent header of go-github) accessed the GitHub API with a token that has an outdated format.
> 
> In order to provide additional security benefits to all our customers, we recently updated the format of our API authentication tokens. Please update your authentication tokens to take advantage of these benefits. You can reset your token with this API endpoint: https://docs.github.com/en/rest/reference/apps#reset-a-token.
> 
> To understand more about this change, and why it's important, visit https://developer.github.com/changes/2021-04-05-behind-githubs-new-authentication-token-formats.
> 
> Thanks,
> The GitHub Team

### PR DESCRIPTION
**Pull request by levb**
Hopefully fixes #431 and https://mattermost.atlassian.net/browse/MM-34646.

Perf. note: this will scan the entire KV set of the plugin every time on startup. There is no "all is done, don't do it again" flag for the scan. @hanzei @catalintomai I don't really see it as an issue, but it's easy to persist the state (in a separate KV key?).

### CODE REVIEW

File: server/plugin/mm_34646_token_refresh.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+func (p *Plugin) forceResetAllMM34646() error {
+	config := p.getConfiguration()
+	ctx := context.Background()
+
+	for page := 0; ; page++ {
+		keys, appErr := p.API.KVList(page, pageSize)
+		if appErr != nil {
+			return appErr
+		}
```
Comment (catalintomai) [inline server/plugin/mm_34646_token_refresh.go:24]: should we continue here (like below)?

File: server/plugin/mm_34646_token_refresh.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+	"encoding/json"
+	"net/http"
+	"time"
+
+	"github.com/google/go-github/v31/github"
+	"github.com/pkg/errors"
+)
+
+const pageSize = 100
+const delayBetweenPages = 5 * time.Minute
```
Comment (catalintomai) [inline server/plugin/mm_34646_token_refresh.go:14]: what do we gain by having such delay? Would it be better to introduce some (smaller) delay for after each (key) reset?

File: server/plugin/mm_34646_token_refresh.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+func (p *Plugin) forceResetAllMM34646() error {
+	config := p.getConfiguration()
+	ctx := context.Background()
+
+	for page := 0; ; page++ {
+		keys, appErr := p.API.KVList(page, pageSize)
+		if appErr != nil {
+			return appErr
```
Comment (catalintomai) [inline server/plugin/mm_34646_token_refresh.go:23]: Should we _continue_ here (like below)?

File: server/plugin/mm_34646_token_refresh.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+	if info.MM34646ResetTokenDone {
+		return info.Token.AccessToken, nil
+	}
+
+	client, err := p.getResetUserTokenMM34646Client(config)
+	if err != nil {
+		p.API.LogInfo("Failed to create a special GitHub client to refresh the user's token", "error", err.Error())
+	}
+
+	req, apiErr := client.NewRequest(http.MethodPatch,
```
Comment (hanzei) [inline server/plugin/mm_34646_token_refresh.go:77]: Can the API method [`Reset`](https://pkg.go.dev/github.com/google/go-github/v38/github#AuthorizationsService.Reset) be used here instead of a custom one?

File: server/plugin/plugin.go [hunk 5]
```diff
@@ -227,6 +252,9 @@ type GitHubUserInfo struct {
 	LastToDoPostAt      int64
 	Settings            *UserSettings
 	AllowedPrivateRepos bool
+
+	// MM34646ResetTokenDone is set for a user whose token has been reset for MM-34646.
+	MM34646ResetTokenDone bool
```
Comment (hanzei) [inline server/plugin/plugin.go:257]: I'm all for linking ticket in code comment for additional context. But naming a variable after it seems to much as it doesn't add clarity to the code. Can this be simply `ResetTokenDone`?

File: server/plugin/mm_34646_token_refresh.go [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+}
+
+func (p *Plugin) forceResetUserTokenMM34646(ctx context.Context, config *Configuration, info GitHubUserInfo) (string, error) {
+	if info.MM34646ResetTokenDone {
+		return info.Token.AccessToken, nil
+	}
+
+	client, err := p.getResetUserTokenMM34646Client(config)
+	if err != nil {
+		p.API.LogInfo("Failed to create a special GitHub client to refresh the user's token", "error", err.Error())
```
Comment (hanzei) [inline server/plugin/mm_34646_token_refresh.go:74]: Should the code `return` here?

Comment (hanzei) [inline server/plugin/mm_34646_token_refresh.go:74]: Debug or Info?

File: server/plugin/plugin.go [hunk 7]
```diff
@@ -84,7 +84,22 @@ func NewPlugin() *Plugin {
 	return p
 }
 
-func (p *Plugin) githubConnect(token oauth2.Token) *github.Client {
+func (p *Plugin) githubConnectUser(ctx context.Context, info *GitHubUserInfo) *github.Client {
+	access := info.Token.AccessToken
+	config := p.getConfiguration()
+	updated, err := p.forceResetUserTokenMM34646(ctx, config, *info)
```
Comment (hanzei) [inline server/plugin/plugin.go:90]: I think the performance impact of this is fine. On a update of the server all connected clients will reset there tokens meaning `#Users` requests are made. But they will be made anyway, as clients tend to refresh after an update.

File: server/plugin/plugin.go [hunk 8]
```diff
@@ -152,6 +171,12 @@ func (p *Plugin) OnActivate() error {
 
 	registerGitHubToUsernameMappingCallback(p.getGitHubToUsernameMapping)
 
+	go func() {
+		err := p.forceResetAllMM34646()
+		if err != nil {
+			p.API.LogDebug("failed to reset user tokens", "error", err.Error())
+		}
```
Comment (hanzei) [inline server/plugin/plugin.go:178]: Why is this needed? Is the lazy migration when the clients connects not enough?

**Review by hanzei — CHANGES_REQUESTED** [review]
Thanks for delivering a programmatic solution @levb!

File: server/plugin/mm_34646_token_refresh.go (hunk 3 above)
Comment (levb) [reply server/plugin/mm_34646_token_refresh.go:22]: 1/5 this is a "bad" error, so we should fail and bail out. Below are mostly expected conditions (like skipping keys we don't understand). I can see the argument for either bailing on KV.Get/Set appErrs below.

File: server/plugin/mm_34646_token_refresh.go (hunk 4 above)
Comment (levb) [reply server/plugin/mm_34646_token_refresh.go]: I couldn't find the basic auth APIs/cli, thx for the pointer (I was only searching for `ResetToken`)

File: server/plugin/plugin.go (hunk 8 above)
Comment (levb) [reply server/plugin/plugin.go:178]: I am not sure if the "old" tokens will still be resettable once they don't support that format; so 1/5 we should proactively reset? Conversely, I am not trysting a 1-time global backfill script to reset all of the tokens, so I put in the lazy user-specific reset as well.

File: server/plugin/plugin.go (hunk 7 above)
Comment (levb) [reply server/plugin/plugin.go:90]: 2/5 Good point (active clients resetting immediately) - We may want to delay the global reset script by a few minutes to get the active users a chance to reset first?

File: server/plugin/plugin.go (hunk 5 above)
Comment (levb) [reply server/plugin/plugin.go:257]: 1/5 I disagree. The names should be intention-revealing, and `ResetTokenDone` is not specific enough

File: server/plugin/mm_34646_token_refresh.go (hunk 6 above)
Comment (levb) [reply server/plugin/mm_34646_token_refresh.go]: should've been neither, but return the error; thanks for catching this one!

File: server/plugin/mm_34646_token_refresh.go (hunk 2 above)
Comment (catalintomai) [reply server/plugin/mm_34646_token_refresh.go:13]: Yeah, I also don't know what the server considers "bad" request policy - blasts coming at larger interval vs. a constant trickle. 0/5 as well, unless we get more context.

File: server/plugin/mm_34646_token_refresh.go (hunk 3 above)
Comment (catalintomai) [reply server/plugin/mm_34646_token_refresh.go:22]: I was approaching this from the angle of a "KV" operation failing for some reason (where KVGet fails for a key we continue, if KVList fails for a page we bail out).  I can see the argument where a page failure is more impactful than a per-key failure (though on the extreme we could get all key failures). 0/5 on this, just wanted to point the diff. handling.

File: server/plugin/mm_34646_token_refresh.go (hunk 2 above)
Comment (levb) [reply server/plugin/mm_34646_token_refresh.go:13]: Based on these comments, https://github.com/mattermost/mattermost-plugin-github/pull/472#discussion_r693777591 and https://github.com/mattermost/mattermost-plugin-github/pull/472#discussion_r693778173 I'll do a 1-time delay of say 5 minutes, and then a slow trickle, 1 at a time. cc @hanzei

File: server/plugin/mm_34646_token_refresh.go (hunk 3 above)
Comment (levb) [reply server/plugin/mm_34646_token_refresh.go:22]: Given the redundant nature of this script, `continue` is fine, will change.

Comment (levb) [reply server/plugin/mm_34646_token_refresh.go:22]: Come to think of it, we'd  return upon `KVList` failure anyway (`len(keys) < pageSize)`) sp keeping as is.

**Review by catalintomai — APPROVED** [review]

File: server/plugin/plugin.go (hunk 7 above)
Comment (hanzei) [reply server/plugin/plugin.go:90]: Good idea :+1:

File: server/plugin/mm_34646_token_refresh.go [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+	time.Sleep(delayToStart)
+
+	for page := 0; ; page++ {
+		keys, appErr := p.API.KVList(page, pageSize)
+		if appErr != nil {
+			return appErr
+		}
+
+		for _, key := range keys {
+			data, appErr := p.API.KVGet(key)
```
Comment (hanzei) [inline server/plugin/mm_34646_token_refresh.go:29]: It's a PITA that there is no user info prefix :/

**Review by hanzei — APPROVED** [review]
Looks good to me. Once we push a release containing this change to community, we should monitor the performance impact.

**Comment by levb** [comment]
@DHaussermann marked you for QA review, but not sure what test steps to take other than watching the logs 5 mins after the plugin start and making sure there's no unexpected errors.

**Review by cpanato — APPROVED** [review]
super nice! thanks for adding this

**Comment by hanzei** [comment]
@levb Do you know any additional methods that can be used to verify that this changes works as expected? If not, I think we can merge this PR and use Community as a canary.

**Comment by levb** [comment]
@hanzei no
